src/runtime/task/mod.rs

```diff
@@ -285,3 +324,55 @@ impl Debug for TaskSet {
         write!(f, "}}")
     }
 }
+
+/// A unique identifier for a [`LocalKey`](crate::thread::LocalKey)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct LocalKeyId(usize);
+
+impl LocalKeyId {
+    fn new<T: 'static>(key: &'static LocalKey<T>) -> Self {
+        Self(key as *const _ as usize)
+    }
+}
+
+/// A map of thread-local storage values. Values are Option<_> because we need to be able to
+/// iteratively destruct them, as it's valid for TLS destructors to initialize new TLS slots.
+struct LocalMap(HashMap<LocalKeyId, Option<Box<dyn Any>>>);
```

Comment:
Since `LocalKeyId` can be ordered (being a `usize`), how about using `BTreeMap` instead?  That way the drop order will be consistent across runs, in case we need reproducibility?

Reply:
Deterministic drop order is a really good point. Deriving `Ord` on `LocalKeyId` isn't quite enough because its values are computed from pointers, which aren't guaranteed to be deterministic across runs. I ended up just biting the bullet and making `pop` return slots in insertion order.